### PR TITLE
DateFormatFixture to do simple date format conversions

### PR DESCRIPTION
**Pull request by tcnh**
Simple utility fixture to convert between different date formats and translate (milli)seconds from epoch to formatted date strings and vice versa.

Also includes a wiki page with some usage examples.

### CODE REVIEW

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public String formatDateAs(String date, String newFormat) {
+        SimpleDateFormat sdf = new SimpleDateFormat(dateFormat);
+        SimpleDateFormat targetFormat = new SimpleDateFormat(newFormat);
+        String formattedDate = null;
+        try {
+            formattedDate = targetFormat.format(sdf.parse(date));
+        } catch (ParseException e) {
+            e.printStackTrace();
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Instead of printing a stack trace to standard error, why not throw a `SlimFixtureException` with message that date can not be parsed and `e` as cause?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        long timeStamp = 0;
+        SimpleDateFormat sdf = new SimpleDateFormat(dateFormat);
+
+        try {
+            timeStamp = sdf.parse(date).getTime();
+            if (!timestampHasMilliseconds) {
+                timeStamp = timeStamp / 1000L;
+            }
+        } catch (ParseException e) {
+            e.printStackTrace();
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Instead of printing a stack trace to standard error, why not throw a `SlimFixtureException` with message that date can not be parsed and `e` as cause?

File: wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+!3 Examples
+
+!define dateTime {!today (dd-MM-yyyy HH:mm:ss)}
+!define date {!today (dd-MM-yyyy)}
+
+|script |date format fixture                             |
+|$stamp=|timestamp for date|${date}                      |
+|show   |format timestamp  |$stamp                       |
+|show   |format timestamp  |$stamp|as|yyyy/MM/dd HH:mm:ss|
+
```
Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki]: Add a check that (with correct format) `$stamp` can be formatted to give `${date}`

File: wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+|script                    |date format fixture                                  |
+|timestamp has milliseconds|false                                                |
+|set date format           |dd-MM-yyyy HH:mm:ss                                  |
+|show                      |format date       |${dateTime}|as|yyyy/MM/dd HH:mm   |
+|$stamp=                   |timestamp for date|${dateTime}                       |
+|show                      |format timestamp  |$stamp                            |
+|show                      |format timestamp  |$stamp     |as|yyyy/MM/dd HH:mm:ss|
+
```
Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki]: Add a check that (with correct format) `$stamp` can be formatted to give ${dateTime}

**Comment by fhoeben** [comment]
Should there be a method to configure the timezone to use when converting timestamp to readable time?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Date;
+import java.util.TimeZone;
+
+/**
+ * Fixture that helps converting dates to and from various formats. Intended to be used as a utility fixture i.e. as library
+ */
+
+public class DateFormatFixture extends SlimFixture {
+    private String dateFormat = "dd-MM-yyyy";
+    private TimeZone timezone = TimeZone.getTimeZone("GMT+1");
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Why not use `TimeZone.getDefault()`, might be more intuitive for people in different parts of the world?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+/**
+ * Fixture that helps converting dates to and from various formats. Intended to be used as a utility fixture i.e. as library
+ */
+
+public class DateFormatFixture extends SlimFixture {
+    private String dateFormat = "dd-MM-yyyy";
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Why not default to `((SimpleDateFormat) DateFormat.getDateInstance(DateFormat.SHORT)).toPattern();`, which uses the Java runtime default locale. For the Netherlands that would get you 'd-M-yy', which means no leading zeroes and year in two digits.
In tests you could/should than explicitly state the format to use, since that is probably not what you want exactly, but it is nicer for users in countries using a different format.

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        } catch (ParseException e) {
+            throw new SlimFixtureException(false, "Could not parse " + date + " using format: " + dateFormat, e);
+        }
+        return formattedDate;
+    }
+
+    public long timestampForDate(String date) {
+        long timeStamp;
+        SimpleDateFormat sdf = new SimpleDateFormat(dateFormat);
+        sdf.setTimeZone(timezone);
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Maybe a method to do the lines below would make the class a bit smaller?
```
        SimpleDateFormat sdf = new SimpleDateFormat(dateFormat);
        sdf.setTimeZone(timezone);
        return sdf;
```

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java (hunk 6 above)
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: I thought about using local defaults for timezone and format but decided to use defaults, beacouse using system settings yields unexpected results. For instance in one of my current projects, most testers create tests on their local machine, that has a dutch locale and a western europe timezone. The Jenkins master is a local server with a US locale and a western europe timezone and some jenkins slaves are EC2 instances with a US locale and a different timezone.

Setting a default ensures similar behaviour wherever the test runs. Your thoughts on that?

Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: This is true. My first impulse is to explicitly set the desired values from the wiki in such a scenario. But in saying that the current solution works just as well, it also allows an explicit set for people that don't like the defaults coded in the class.

BUT, on the other hand, it is hard to back to using the system locale (as far as I know there is no string you can pass to explicitly use the system default, you have to use the constants from code).

So my preference is still to have the system defaults in code, and use these as defaults. This will be more intuitive to 'first-time' users. Where 'power-users' like yourself can set explicit values to ensure the test runs the same on any environment.

You could consider adding an extra constructor which explicitly configures the format and timezone, so you can still [use the fixture and configure it in a single table row](http://fitnesse.org/FitNesse.FullReferenceGuide.UserGuide.WritingAcceptanceTests.SliM.ConstructorArguments) (instead of needing 3).

What do you think?

**Comment by fhoeben** [comment]
Did you notice https://github.com/fhoeben/hsac-fitnesse-fixtures/tree/enterDateField? I think that will greatly benefit from this fixture, so the user can ensure the date they want to enter is in yyyy-mm-dd format.
Do you have any comments/thoughts?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java (hunk 6 above)
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Agree. I like the extra constructor option so I'll implement that and set defaults to system default values.

File: wiki/FitNesseRoot/FrontPage/content.txt [hunk 8]
```diff
@@ -13,3 +13,6 @@ To add your first "page", click the [[Edit][.FrontPage?edit]] button and add a [
 |[[Release Notes][.FitNesse.ReleaseNotes]]                     |''Find out about FitNesse's new features'' |
 
 !note Release ${FITNESSE_VERSION}
+
+
+${prop}
```
Comment (fhoeben) [inline wiki/FitNesseRoot/FrontPage/content.txt]: Why this change?

Comment (tcnh) [inline wiki/FitNesseRoot/FrontPage/content.txt]: Should not have been committed.

**Comment by fhoeben** [comment]
I refactored a bit. Do you agree?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    protected Date parseDate(SimpleDateFormat sdf, String date) {
+        try {
+            return sdf.parse(date);
+        } catch (ParseException e) {
+            String pattern = sdf.toPattern();
+            throw new SlimFixtureException(false, "Could not parse " + date + " using format: " + pattern, e);
+        }
+    }
+
+    protected SimpleDateFormat getConfiguredDateFormat() {
```
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Maybe make this method public?
Use case: use the system default format in other tests as well (this may also help in converting for BrowserTest's new enterDate method?)

File: wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki [hunk 10]
```diff
@@ -45,18 +45,33 @@ This fixture is intended to be used as a library. (See below)
...
+By default a library will use system default date format and timezone.
+
+|library            |
+|date format fixture|
+
+Example usage of date formatting methods from a string fixture script table:
+
+|script |string fixture                                  |
+|$stamp=|timestamp for date|${date}                      |
```
Comment (tcnh) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki]: If system defaults are used, we need to specify the input format if we want to ensure passing travis builds

Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki]: You are soooo right (blush)...

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java (hunk 9 above)
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: Sorry, I don't immediately grasp the use case yet. Why would one need to get the system default date format to get a value in yyyy-mm-dd?
One can just use a date format fixture instance to go from whatever format the date is in to yyyy-mm-dd, or ...?

File: wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki (hunk 10 above)
Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/UtilityFixtures/DateFormatFixture.wiki]: That took me more effort than I expected..... ;-)

File: src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java (hunk 9 above)
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/DateFormatFixture.java]: You're right. I was thinking of situations where one would get a date from a SUT, but the locale of the SUT has no relation to the locale of the machine running the test :)

Being able to use the date format fixture as a library should suffice for all date related use cases
